akka-persistence/src/main/scala/akka/persistence/PersistentActor.scala

```diff
@@ -99,6 +101,58 @@ object Recovery {
 }
 
 /**
+ * This defines how to handle the current received message which failed to stash, when the size of internal
+ * Stash exceeding the capacity of Stash.
+ */
+sealed trait StashOverflowStrategy {
+  def handleFailure(message: Any, sender: ActorRef, original: StashOverflowException)
+                   (implicit context: ActorContext, receiver: ActorRef)
+}
+
+/**
+ * Discard the message to [[DeadLetter]].
+ */
+case object DiscardToDeadLetter extends StashOverflowStrategy {
+  /**
+   * Java API: get the singleton instance
+   */
+  def getInstance = this
+
+  override def handleFailure(message: Any, sender: ActorRef, original: StashOverflowException)
+                            (implicit context: ActorContext, receiver: ActorRef): Unit = {
+    context.system.deadLetters.tell(DeadLetter(message, sender, receiver), sender)
```

Comment:
I understand that you used `ActorContext` here to be able to grab the implicit from the enclosing actor, but I think we should not pass the ActorContext around, even though it's only to our own implementations. 

Either change the param to `system`, or perhaps better remove the abstract `handleFailure` method and handle the implementation with a pattern match at the call site instead.